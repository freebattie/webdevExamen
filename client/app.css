*{
    margin: 0;
    padding: 0;
    border: 0;
    box-sizing: border-box;

    font-family: 'Fira Sans', sans-serif;
}

.menu{
    margin-top: 40px;
    padding: 100px;
    display: flex;

    border: 1px solid black;
    flex-wrap: wrap;
    justify-content: center;

}
.dish{
    margin: 10px;
    width: 300px;
    border: solid 2px black;
    height: 200px;
}
footer{
    width: 80px;
}
.editItem{
    display: flex;
    width: 300px;
    flex-wrap: wrap;
}
.button{
    color: black;
    padding: 5px 2px;
    text-align: center;
    text-decoration: none;
    display: inline-block;
    cursor:pointer;
    margin: 10px;
    font-size: 16px;
    background-color: white;

    border: 2px solid black;
    width: 70px;
    border-radius: 5px;
    transition-duration: 0.4s;
}
.menuheader{
    margin-bottom: 20px;
}
.orders{
    display: flex;
    flex-wrap: wrap;
}
.order{

    height: 300px;
    border: 2px black solid;
    width: 300px;
    margin: 10px;
}
.items{
    width: 100px;
    margin: 10px;
}
.ordersheader{
    width: 100%;
}
.editform{
    width: 400px;
    border: 2px solid black;
}
.itemsToEdit{
    width: 100%;
    display: flex;
    flex-wrap: wrap;

}