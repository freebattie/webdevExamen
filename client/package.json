{
  "name": "client",
  "version": "1.0.0",
  "description": "",
  "type": "module",
  "scripts": {
    "test": "jest --coverage",
    "format": "prettier --write *.{jsx,js,tsx,ts,html}",
    "dev": "parcel index.html",
    "build": "npm install --include=dev && npm run build:parcel",
    "build:parcel": "parcel build index.html"
  },

  "keywords": [],
  "author": "",
  "license": "ISC",
  "devDependencies": {
    "@babel/preset-env": "^7.19.1",
    "@babel/preset-react": "7.18.6",
    "@types/react": "^18.0.21",
    "@types/react-dom": "^18.0.6",
    "babel-jest": "^29.0.3",
    "events": "^3.3.0",
    "jest": "^29.0.3",
    "jest-environment-jsdom": "^29.0.3",
    "parcel": "^2.7.0",
    "process": "^0.11.10",
    "rimraf": "^3.0.2"
  },
  "dependencies": {
    "@parcel/transformer-react-refresh-wrap": "^2.7.0",
    "@types/jest": "^29.0.3",
    "enzyme": "^3.11.0",
    "history": "^5.3.0",
    "react": "^18.2.0",
    "react-dom": "^18.2.0",
    "react-router-dom": "^6.4.0"
  },
  "jest": {
    "testEnvironment": "jsdom",
    "collectCoverageFrom": [
      "**/*.jsx",
      "**/*.js",
      "!dist/**",
      "!coverage/**",
      "!app.jsx",
      "!index.jsx",
      "!lib/**"
    ]
  },
  "babel": {
    "presets": [
      "@babel/preset-env",
      [
        "@babel/preset-react",
        {
          "runtime": "automatic"
        }
      ]
    ]
  }
}
