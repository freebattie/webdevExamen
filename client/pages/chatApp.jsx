import React, { useState } from "react";
import { useNavigate } from "react-router-dom";

export function ChatApplication({ messages, onNewMessage }) {
  const [message, setMessage] = useState("");
  const navigate = useNavigate();
  function handleSubmit(e) {
    e.preventDefault();
    onNewMessage(message);
    setMessage("");
  }
  return (
    <>
      <header className={"footermove headerText"}>Modern Snack Chat</header>
      <main className={" border"}>
        {messages.map(({ message, user }, index) => (
          <div key={index}>
            <strong>{user}:</strong> {message}
          </div>
        ))}
      </main>
      <footer className={"footermove"}>
        <form className={"formsize"} onSubmit={handleSubmit}>
          <input
            className={"input2"}
            autoFocus={true}
            value={message}
            onChange={(e) => setMessage(e.target.value)}
          />
          <button className={"button"}>Send</button>
          <button className={"button"} onClick={(e) => navigate("/")}>
            back
          </button>
        </form>
      </footer>
    </>
  );
}
