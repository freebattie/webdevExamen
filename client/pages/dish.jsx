import { useEffect, useState } from "react";

export function Dish(props) {
  const [number, setNumber] = useState();
  let nextId = number;
  useEffect(() => {
    setNumber(props.val);
    const old = [...props.orders];
    old[0] = null;
    props.setOrders(old);
  }, []);
  async function handelAddToList() {
    setNumber((perv) => perv + 1);
    nextId++;
    setNumber(nextId);
    const item = {
      id: props.d.id,
      name: props.d.name,
      description: props.d.description,

      type: props.d.type,
      price: props.d.price,
      size: nextId,
    };
    const old = [...props.orders];
    old[props.index] = item;
    props.setOrders(old);
  }
  async function handelRemoveToList() {
    if (number > 0) {
      nextId--;
      setNumber(nextId);
      const item = {
        id: props.d.id,
        name: props.d.name,
        price: props.d.price,
        size: nextId,
      };
      const old = [...props.orders];
      old[props.d.id] = item;
      props.setOrders(old);
    }
  }

  return (
    <div className={"dish"}>
      <div>name :{props.d.name}</div>
      <div>price:{props.d.price}Nok</div>
      <div>description:{props.d.description}</div>

      <div>type Dish :{props.d.type}</div>
      <div>order :{number}</div>
      <button className={"button"} onClick={() => handelAddToList()}>
        add
      </button>
      <footer>
        <button className={"button"} onClick={() => handelRemoveToList()}>
          remove
        </button>
      </footer>
    </div>
  );
}
