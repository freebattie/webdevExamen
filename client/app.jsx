import { BrowserRouter, Routes, Route } from "react-router-dom";
import { useState } from "react";
import { FrontPage } from "./pages/frontpage.jsx";
import { EditUsers } from "./pages/edituser.jsx";
import { ErrorUser } from "./pages/erroruser.jsx";
import { Login } from "./pages/login.jsx";
import { NewUser } from "./pages/newuser.jsx";
import "./app.css";
function Error(props) {
  return (
    <div style={{ border: "1px solid red", background: "Pink" }}>
      An error occurred: {props.error.toString()}
    </div>
  );
}

export function App() {
  const [error, setError] = useState();
  return (
    <BrowserRouter>
      <Routes>
        <Route path={"/"} element={<FrontPage />} />
        <Route path={"/login"} element={<Login setError={setError} />} />
        <Route path={"/register"} element={<NewUser setError={setError} />} />
        <Route path={"/edit"} element={<EditUsers />} />
        <Route path={"/error"} element={<Error error={error} />} />
        <Route path={"/errorUser"} element={<ErrorUser />} />
      </Routes>
    </BrowserRouter>
  );
}
