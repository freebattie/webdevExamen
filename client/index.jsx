import { createRoot } from "react-dom/client";

import { App } from "./app";

// get the root element in index.html
const container = document.getElementById("app");
//creat a react root from that id
const root = createRoot(container);

root.render(<App />);
