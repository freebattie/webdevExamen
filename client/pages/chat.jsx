import { useEffect, useState } from "react";
import { ChatApplication } from "./chatApp.jsx";
const initialMessages = [];

function UserRegistrationForm({ onUsername }) {
  const [username, setUsername] = useState("");
  function handleSubmit(event) {
    event.preventDefault();
    onUsername(username);
  }
  return (
    <center>
      <form className={"loginChat"} onSubmit={handleSubmit}>
        <div className={"loginChatfld"}>
          <label>
            Username:
            <input
              className={"input"}
              value={username}
              onChange={(e) => setUsername(e.target.value)}
            />
          </label>
          <button className={"button"}>Submit</button>
        </div>
      </form>
    </center>
  );
}
export function Chat() {
  const [user, setUser] = useState();
  const [messages, setMessages] = useState(initialMessages);
  const [ws, setWs] = useState();

  useEffect(() => {
    const ws = new WebSocket("ws://" + window.location.host + "/api/chat");
    ws.onopen = (event) => {
      console.log("Opened", event);
    };
    ws.onmessage = (event) => {
      console.log("message", event);
      const { user, message } = JSON.parse(event.data);
      setMessages((messages) => [...messages, { message, user }]);
    };
    setWs(ws);
  }, []);

  function handleNewMessage(message) {
    ws.send(JSON.stringify({ message, user }));
  }

  if (!user) {
    return <UserRegistrationForm onUsername={setUser} />;
  }

  return (
    <ChatApplication messages={messages} onNewMessage={handleNewMessage} />
  );
}
