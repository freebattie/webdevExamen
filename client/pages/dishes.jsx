import { useContext, useState } from "react";
import { Appcontext } from "../lib/appcontext.jsx";
import { useNavigate } from "react-router-dom";
import { useLoader } from "../lib/useLoading.jsx";
import { ErrorMsg } from "./errormsg.jsx";
import { Dish } from "./dish.jsx";
import { addGivenOrder } from "../lib/addgivenorder.jsx";

export function Dishes({ setError }) {
  const { listDishes } = useContext(Appcontext);
  const [orders, setOrders] = useState([{ price: 0, size: 0 }]);
  const [time, setTime] = useState("08:00");
  const [date, setDate] = useState("");
  const [location, setLocation] = useState("");
  const navigate = useNavigate();
  const { loading, error, data, reload } = useLoader(
    async () => await listDishes()
  );
  const { addOrder } = useContext(Appcontext);

  async function handelAddOrder(e) {
    e.preventDefault();

    const val = [...orders];
    console.log(val);
    let max = 0;
    val.forEach((e) => {
      if (e != null) {
        max += e.price * e.size;
      } else {
      }
    });

    const newOrder = {
      location,
      date: date.length < 1 ? today : date,
      time,
      orders,
      total: max,
    };
    await addGivenOrder(newOrder, setError, navigate, addOrder);
  }
  if (error) {
    return <ErrorMsg error={error} />;
  }
  if (loading) {
    return <h1>Loading....</h1>;
  }
  function getDate() {
    let today = new Date();
    let dd = today.getDate();
    let mm = today.getMonth() + 1; //January is 0!
    let yyyy = today.getFullYear();

    if (dd < 10) {
      dd = "0" + dd;
    }

    if (mm < 10) {
      mm = "0" + mm;
    }

    return yyyy + "-" + mm + "-" + dd;
  }

  let today = getDate();

  return (
    <center>
      <div className={"menuheader"}>DISHES:</div>
      <div className={"menu"}>
        {data.map((d, index) => {
          return (
            <Dish
              key={d.id}
              d={d}
              setOrders={setOrders}
              orders={orders}
              val={0}
              index={index}
            />
          );
        })}
      </div>
      <form onSubmit={(e) => handelAddOrder(e)}>
        <div>
          <label>
            Location:
            <input
              className={"input"}
              type="text"
              value={location}
              onChange={(e) => setLocation(e.target.value)}
            />
          </label>
          <label>
            date:
            <input
              className={"input"}
              type="date"
              value={date}
              min={today}
              onChange={(e) => setDate(e.target.value)}
            />
          </label>
          <label>
            time:
            <input
              className={"input"}
              type="time"
              value={time}
              min={"08:00"}
              max={"21:00"}
              onChange={(e) => setTime(e.target.value)}
            />
          </label>
        </div>
        <button data-testid={"button"} className={"button"}>
          Order
        </button>
      </form>
    </center>
  );
}
