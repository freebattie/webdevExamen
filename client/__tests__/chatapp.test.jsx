import * as React from "react";

import { createRoot } from "react-dom/client";
import { act, Simulate } from "react-dom/test-utils";
import { MemoryRouter } from "react-router-dom";
import { ChatApplication } from "../pages/chatApp";

global.IS_REACT_ACT_ENVIRONMENT = true;

describe("app", () => {
  it("should rener App", async function () {
    const element = document.createElement("div");

    //creat a React root from that id
    const root = createRoot(element);
    const onNewMessage = jest.fn();
    await act(async () => {
      root.render(
        <MemoryRouter>
          <ChatApplication
            messages={["en", "to"]}
            onNewMessage={onNewMessage}
          />
        </MemoryRouter>
      );
    });
    await act(async () => {
      await Simulate.change(await element.querySelector("footer form input"), {
        target: { value: "Petter petter" },
      });
    });
    await act(async () => {
      await Simulate.submit(await element.querySelector("form"));
    });
    expect(element).toMatchSnapshot();
    await act(async () => {});
  });
});
