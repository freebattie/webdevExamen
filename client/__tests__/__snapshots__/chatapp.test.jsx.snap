// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app should rener App 1`] = `
<div>
  <header
    class="footermove headerText"
  >
    Modern Snack Chat
  </header>
  <main
    class=" border"
  >
    <div>
      <strong>
        :
      </strong>
       
    </div>
    <div>
      <strong>
        :
      </strong>
       
    </div>
  </main>
  <footer
    class="footermove"
  >
    <form
      class="formsize"
    >
      <input
        class="input2"
        value=""
      />
      <button
        class="button"
      >
        Send
      </button>
      <button
        class="button"
      >
        back
      </button>
    </form>
  </footer>
</div>
`;
